content/creating-addons/react/en/decorators.md

```diff
@@ -0,0 +1,438 @@
+
+import { clearStyles, addOutlineStyles } from './helpers';
+import outlineCSS from './outlineCSS';
+
+export const withGlobals = (StoryFn, context) => {
+  const [{ outlineActive }, updateGlobals] = useGlobals();
+  // Is the addon being used in the docs panel
+  const isInDocs = context.viewMode === 'docs';
+
+  const outlineStyles = useMemo(() => {
+    const selector = isInDocs ? `#anchor--${context.id} .docs-story` : '.sb-show-main';
+
+    return outlineCSS(selector);
+  }, [context.id]);
+
+  useEffect(() => {
+    const selectorId = isInDocs ? `addon-outline-docs-${context.id}` : `addon-outline`;
+
+    if (!isActive) {
+      clearStyles(selectorId);
+      return;
+    }
+
+    addOutlineStyles(selectorId, outlineStyles);
+
+    return () => {
+      clearStyles(selectorId);
+    };
+  }, [isActive, outlineStyles, context.id]);
+};
+```
+
+Ok, that seems like a big jump. Let’s walk through all the changes.
+
+The addon can be active in both docs and story view modes. The actual DOM node for the preview `iframe` is different in these two modes. In fact, the docs mode renders multiple story previews on one page. Therefore, we need to pick the appropriate selector for the DOM node where the styles will be injected. Also, the CSS needs to be scoped to that particular selector.
+
+<div class="aside"><b>Note:</b> the <code>useMemo</code> and <code>useEffect</code> here come from <a href="https://storybook.js.org/docs/react/addons/addons-api">@storybook/addons</a> and not React. This is because the decorator code is running in the preview part of Storybook. That's where the user's code is loaded which may not contain React. Therefore, to be framework agnostic, Storybook implements a React-like hook library which we can use!</div>
+
+Next, as we inject the styles into the DOM, we need to keep track of them to clear them when the user toggles it off or switches the view mode.
+
+To manage all this CSS logic, we need a few helpers. These use DOM APIs to inject and remove stylesheets.
+
+```js:title=src/helpers.js
+/* eslint-env browser */
+export const clearStyles = selector => {
+  const selectors = Array.isArray(selector) ? selector : [selector];
+  selectors.forEach(clearStyle);
+};
+
+const clearStyle = selector => {
+  const element = document.getElementById(selector);
+  if (element && element.parentElement) {
+    element.parentElement.removeChild(element);
+  }
+};
+
+export const addOutlineStyles = (selector, css) => {
+  const existingStyle = document.getElementById(selector);
+  if (existingStyle) {
+    if (existingStyle.innerHTML !== css) {
+      existingStyle.innerHTML = css;
+    }
+  } else {
+    const style = document.createElement('style');
+    style.setAttribute('id', selector);
+    style.innerHTML = css;
+    document.head.appendChild(style);
+  }
+};
+```
+
+And the outline CSS itself is based on what [Pesticide uses](https://github.com/mrmrs/pesticide/blob/master/css/pesticide.css).
+
+```js:title=src/outlineCSS.js
```

Comment:
Can we make this CSS more compact? It's a lot of scrolling.

If not, I'd suggest we link to a GitHub file where they can download the CSS.